GitUI/Theming/ThemeCssLoader.cs

```diff
@@ -0,0 +1,216 @@
+
+        private readonly TranslationString _invalidRule = new TranslationString("Invalid rule: {0}");
+        private readonly TranslationString _failedToLoadThemeFrom = new TranslationString("Failed to load theme from {0}");
+        private readonly TranslationString _fileNotFound = new TranslationString("File not found");
+        private readonly TranslationString _fileTooLarge = new TranslationString("File too large");
+
+        private readonly Parser _parser;
+        private readonly ICssUrlResolver _urlResolver;
+        private readonly string[] _allowedClasses;
+
+        private readonly Dictionary<AppColor, Color> _appColors = new Dictionary<AppColor, Color>();
+        private readonly Dictionary<KnownColor, Color> _sysColors = new Dictionary<KnownColor, Color>();
+        private readonly Dictionary<string, int> _specificityByColor = new Dictionary<string, int>();
+
+        private bool _parseCalled;
+
+        public ThemeCssLoader(ICssUrlResolver urlResolver, string[] allowedClasses)
+        {
+            _parser = new Parser();
+            _urlResolver = urlResolver;
+            _allowedClasses = allowedClasses;
+        }
+
+        public IReadOnlyDictionary<AppColor, Color> AppColors => _appColors;
+
+        public IReadOnlyDictionary<KnownColor, Color> SysColors => _sysColors;
+
+        public bool TryLoadCss(string filePath)
+        {
+            if (_parseCalled)
+            {
+                throw new InvalidOperationException($"{nameof(ThemeCssLoader)} only supports 1 call to {nameof(TryLoadCss)}");
+            }
+
+            _parseCalled = true;
+
+            return TryLoadCssImpl(filePath, cssImportChain: Array.Empty<string>());
+        }
+
+        private bool TryLoadCssImpl(string filePath, string[] cssImportChain)
+        {
+            if (!TryReadFile(filePath, out string inputContent))
+            {
+                return false;
+            }
+
+            var stylesheet = _parser.Parse(inputContent);
+            foreach (var importDirective in stylesheet.ImportDirectives)
+            {
+                if (!TryImport(filePath, importDirective, cssImportChain))
+                {
+                    return false;
+                }
+            }
+
+            foreach (StyleRule rule in stylesheet.StyleRules)
+            {
+                if (!TryParseRule(filePath, rule))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private bool TryReadFile(string fileName, out string result)
+        {
+            result = null;
+            var fileInfo = new FileInfo(fileName);
+
+            if (!fileInfo.Exists)
+            {
+                PrintTraceWarning(fileName, _fileNotFound.Text);
+                return false;
+            }
+
+            if (fileInfo.Length > 1024 * 1024)
+            {
+                PrintTraceWarning(fileName, _fileTooLarge.Text);
```

Comment:
Overall what was the reason to use this approach instead of throwing?
E.g.
```cs
        private string ReadThemeFile(string themeFileName)
        {
            if (fileInfo.Length > 1024 * 1024)
            {
                throw new FileLoadException(_fileTooLarge.Text, themeFileName);
            }

            return File.ReadAllText(themeFileName);
        }
```

Reply:
Because the theme is loaded when starting up the application, there is no color settings page to display the error yet.

Reply:
- [ ] load theme when user selects it so that if exception happens, it can be shown before restarting the app

Reply:
I'd rather have the app fail with a popup message telling the user there is a problem with a theme (and how to fix it preferably) than silently swallow errors and have the user come here complain about why changes to a theme didn't work.

Reply:
These errors would be user-induced, so the user would be in a position to fix them.

Reply:
ok